notebooks/automl.clj

```diff
@@ -0,0 +1,218 @@
+;;
+;; The ctx acts a a collector of everything "learned" during :fit,
+;; mainly the trained model, but it could be as well other information
+;; learned from the data.
+
+(def train-ctx
+  (mm/fit titanic
+          (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+
+;; (The dummy model does not have state....)
+
+train-ctx
+(defn transform-pipe [ctx data pipe-fn]
+  (mm/transform-pipe data pipe-fn ctx))
+
+;;  Lets prepare data for some examples.
+(require '[tablecloth.api :as tc])
+
+(def splits (first (tc/split->seq titanic)))
+(def train-ds (:train splits))
+(def test-ds (:test splits))
+
+
+;; To show the power of pipelines, I start with doing the simplest possible pipeline,
+;; and expand then on it.
+
+;;  we can already chain train and test with usual functions:
+(->>
+ (ml/train train-ds {:model-type :metamorph.ml/dummy-classifier})
+ (ml/predict test-ds)
+ :survived)
+
+;;  the same with pipelines
+(def pipeline
+  (mm/pipeline (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline)
+ (mm/transform-pipe test-ds pipeline)
+ :metamorph/data :survived)
+
+
+;;  but know we can add more operations to the pipeline,
+;;  and nothing else changes, for example drop columns.
+(def pipeline-2
+  (mm/pipeline
+   (mm/lift tc/drop-columns [:embarked])
+   (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline-2)
+ (mm/transform-pipe test-ds pipeline-2)
+ :metamorph/data :survived)
+
+;; The 'lift` function transposes a dataset->dataset function into a ctx->ctx function, as required for
+;; metamorph pipeline operations
+;;
+;; So a metamorph pipeline can encapsulate arbitray transformation
+;; of a dataset in the 2 modes. They can be "stateless"
+;; (only chaining the dataset, such as `drop-columns`) or
+;; "state-full", so they store data in the ctx during :fit and can use
+;; it in :transform. In the pipeline above, the trained model is
+;; stored in this way.
+;;
+;;This makes pipeline execution "isolated"
+;;
+;; For convenience `tablecloth` contains a ns where all functions
+;; are lifted into ctx->ctx operations, so can be added to pipelines
+;; directly.
+(require 'tablecloth.pipeline)
+(def pipeline-3
+  (mm/pipeline
+   (tablecloth.pipeline/drop-columns [:embarked])
+   (ml/model {:model-type :metamorph.ml/dummy-classifier})))
+(->>
+ (mm/fit-pipe train-ds pipeline-3)
+ (mm/transform-pipe test-ds pipeline-3)
+ :metamorph/data :survived)
+
+;;  It supports as well "pipelines as data"
+(def pipeline-4
+  (mm/->pipeline
+   [[tablecloth.pipeline/drop-columns [:embarked]]]
+   [[ml/model {:model-type :metamorph.ml/dummy-classifier}]]))
+(->>
+ (mm/fit-pipe train-ds pipeline-4)
+ (mm/transform-pipe test-ds pipeline-4)
+ :metamorph/data :survived)
+
+;; The auto ml support in metamorph consists now in the possibility
+;; to create an arbitrary number of different pipelines
+;; and combine them with arbitray test/train data splits
+
```

Comment:
Currently, space lines need to begin with semicolons so that the markdown is rendered correctly.

This is a Clay bug that should be fixed soon.

For now, it is recommended to write
```clj
;;
;; helper for later
```
instead of
```clj

;; helper for later
```